File: server/src/reasoning/mirror.mjs
// The reasoning mirror — the load-bearing half of the splice distillation loop
// (invariant L2). Reasoning items are never replayed upstream (L1, locked
// non-goal); instead each turn's summary is mirrored into the transcript as
// visible text, so conclusions persist while reasoning is re-derived fresh.
//
// ONE mirrorInto() serves both the stream and non-stream paths (v29 duplicated
// the logic at :1616 and :1702 and the copies drifted). Both entry paths pass a
// sink; a permanent test asserts both call it.

// Named thresholds — the single home for every magic number in this loop (P2).
export const MIRROR_MIN_CHARS = 20;   // mirror only when the summary is substantive
export const PROMOTE_MIN_CHARS = 40;  // thinking may be promoted to text at/above this
export const HONESTY_MIN_CHARS = 20;  // completed-but-empty below this → honest error, not blank end_turn

/** WIRE CONTRACT (external): the mirror block format v25–v29 shipped. The
 * any transcript tooling key on this shape. */
export function mirrorWireText(thinking) {
  return `\n[reasoning summary]\n${String(thinking).trim()}\n`;
}

/**
 * Mirror the turn's thinking into the transcript as a visible text block.
 * sink: { addTextBlock(text) } — stream path appends an SSE text block,
 * non-stream path pushes onto the response content array.
 * Returns true when the mirror was emitted.
 */
export function mirrorInto(sink, thinkingText, { showReasoning, compact }) {
  if (compact) return false;              // compact is a text-only summarizer turn
  if (showReasoning !== 'text') return false;
  const t = String(thinkingText || '').trim();
  if (t.length < MIRROR_MIN_CHARS) return false;
  sink.addTextBlock(mirrorWireText(t));
  return true;
}

/** Join the thinking blocks of an Anthropic content array (non-stream path). */
export function extractThinking(content) {
  return (Array.isArray(content) ? content : [])
    .filter((c) => c?.type === 'thinking')
    .map((c) => String(c.thinking || '').trim())
    .filter(Boolean)
    .join('\n\n');
}

File: gateway/gateway/src/main/kotlin/splice/gateway/reasoning/Mirror.kt
// PORT-OF: server/src/reasoning/mirror.mjs @ 4ca99f7 — invariants (L2): ONE mirrorInto serves
// stream and non-stream paths (v29 had two drifting copies; an ast-grep wall pins the
// definition to THIS file); wire contract v25-v29: "\n[reasoning summary]\n<trimmed>\n" —
// transcript tooling keys on this shape; gates: never on compact, only showReasoning=='text',
// only when the trimmed summary >= MIRROR_MIN_CHARS. Thresholds live in :core (shared with
// the dialect's promote step).
package splice.gateway.reasoning

import splice.core.turn.MIRROR_MIN_CHARS
import splice.core.turn.ReasoningDisplay
import splice.core.wire.ContentBlock
import splice.core.wire.ThinkingBlock
import splice.spi.WireSink

/** WIRE CONTRACT (external): the mirror block format transcript tooling keys on. */
public fun mirrorWireText(thinking: String): String = "\n[reasoning summary]\n${thinking.trim()}\n"

/**
 * Mirror the turn's thinking into the transcript as a visible text block.
 * Returns true when the mirror was emitted.
 */
public suspend fun mirrorInto(
    sink: WireSink,
    thinkingText: String?,
    showReasoning: ReasoningDisplay,
    compact: Boolean,
): Boolean {
    if (compact) return false // compact is a text-only summarizer turn
    // Gate cascade (ported contract): only when reasoning is shown as text AND the trimmed
    // summary clears the wire threshold; order preserved (showReasoning gate before length gate).
    val t = thinkingText.orEmpty().trim()
    if (showReasoning != ReasoningDisplay.TEXT || t.length < MIRROR_MIN_CHARS) return false
    sink.addTextBlock(mirrorWireText(t))
    return true
}

/** Join the thinking blocks of an Anthropic content list (non-stream path). */
public fun extractThinking(content: List<ContentBlock>): String =
    content.filterIsInstance<ThinkingBlock>()
        .map { it.thinking.trim() }
        .filter { it.isNotEmpty() }
        .joinToString("\n\n")
